service/storage_service.cc

```diff
@@ -6656,6 +6656,15 @@ future<join_node_response_result> storage_service::join_node_response_handler(jo
 
     co_return co_await std::visit(overloaded_functor {
         [&] (const join_node_response_params::accepted& acc) -> future<join_node_response_result> {
+            // Allow other nodes to mark the replacing node as alive. It has
```

Comment:
I don't understand this comment. Is it one more place where we have to trust gossip UP/DOWN states because we have no tracking of UP/DOWN states of our own? @gleb-cloudius 

What problem are we solving here?

Reply:
I disagree with the purpose: why the heck we need to even look at the gossip alive state of a node when assembling the cluster?

Reply:
> I don't understand this comment. Is it one more place where we have to trust gossip UP/DOWN states because we have no tracking of UP/DOWN states of our own? @gleb-cloudius
>
> What problem are we solving here?

This commit belongs to #15926. I mentioned it in the description.

The problem it solves is described in the commit message:
```
After this change, other nodes can see the replacing node as alive
only after the topology coordinator accepts its join request.

In the following commits, we make the topology coordinator reject
join requests if the node being replaced is considered alive by the
gossiper. However, the replacing node may become alive before the
topology coordinator does the validation. If the replacing node
reuses the IP of the node being replaced, the topology coordinator
cannot know which of these two nodes is alive and whether it should
reject the join request.
```
We don't trust gossip here. It's the other way around -- the gossip state depends on the topology coordinator accepting the join request.

> I disagree with the purpose: why the heck we need to even look at the gossip alive state of a node when assembling the cluster?

We don't need to. In #15926, we reject replace if the node being replaced is considered alive to improve user experience. This change is not entirely safe, but it will be once we implement banning and ban the node being replaced and the ignored nodes at the beginning of replace. It's discussed in https://github.com/scylladb/scylladb/pull/15926#discussion_r1399354469.

Reply:
> I don't understand this comment. Is it one more place where we have to trust gossip UP/DOWN states because we have no tracking of UP/DOWN states of our own? @gleb-cloudius
> 

I have no idea what *you* mean by "tracking of UP/DOWN states of our own". We have the tracking. It is a gossiper's failure detector that we use here. There also not problem trusting UP state. DOWN state is a different story and this is what we discussed with @kbr-scylla here https://github.com/scylladb/scylladb/pull/15926#discussion_r1399362458

> What problem are we solving here?

The UI problem. Currently it is not allowed to remove a live node (you get `Startup failed: std::runtime_error (Cannot replace a live node...)` and it should stay so. It makes a lot of sense. An operator error should not kill a node.